File: apps/api/internal/types/response.go
package types

import "github.com/suprimkhatri77/uptime-monitor/api/internal/constants"

type AppError struct {
	Code    string `json:"code"`
	Field   string `json:"field,omitempty"`
	Message string `json:"message"`
}

type APIResponse struct {
	Success bool            `json:"success"`
	Message string          `json:"message"`
	Code    string          `json:"code,omitempty"`
	Errors  []AppError      `json:"errors,omitempty"`
	Data    any             `json:"data,omitempty"`
	Meta    *PaginationMeta `json:"meta,omitempty"`
}

type PaginationMeta struct {
	Total      int `json:"total"`
	TotalPages int `json:"total_pages"`
	Page       int `json:"page"`
	Limit      int `json:"limit"`
	Offset     int `json:"offset"`
}

func Success(message string, data any) APIResponse {
	resp := APIResponse{
		Success: true,
		Message: message,
	}
	if data != nil {
		resp.Data = data
	}
	return resp
}

func SuccessWithMeta(message string, data any, meta PaginationMeta) APIResponse {
	return APIResponse{
		Success: true,
		Message: message,
		Data:    data,
		Meta:    &meta,
	}
}

func Error(message, code string) APIResponse {
	return APIResponse{
		Success: false,
		Message: message,
		Code:    code,
	}
}

func ValidationErrorResponse(message string, errors []AppError) APIResponse {
	return APIResponse{
		Success: false,
		Message: message,
		Code:    constants.ValidationFailed,
		Errors:  errors,
	}
}

File: packages/types/src/api/response.ts
import { z } from "zod";
import { ErrorCodeSchema } from "./error-codes.js";

export const ValidationErrorSchema = z.object({
  code: z.string(),
  field: z.string().optional(),
  message: z.string(),
});

export type ValidationError = z.infer<typeof ValidationErrorSchema>;

export const apiErrorResponseSchema = z.object({
  success: z.literal(false),
  message: z.string(),
  code: z.string(),
  errors: z.array(ValidationErrorSchema).optional(),
});
export type ApiErrorResponseSchema = z.infer<typeof apiErrorResponseSchema>;
export type ApiErrorResponse = z.infer<typeof apiErrorResponseSchema>;

export type ApiResponse<T = unknown> = {
  success: boolean;
  message?: string;
  code?: z.infer<typeof ErrorCodeSchema>;
  errors?: ValidationError[];
  data?: T;
  meta?: PaginationMeta;
};

export const PaginationMetaSchema = z.object({
  total: z.number(),
  page: z.number(),
  limit: z.number(),
  total_pages: z.number(),
  offset: z.number(),
});

export type PaginationMeta = z.infer<typeof PaginationMetaSchema>;

export function createSuccessResponseSchema() {
  return z.object({
    success: z.literal(true),
    message: z.string(),
  });
}

export function createObjectResponseSchema<T extends z.ZodTypeAny>(
  dataSchema: T,
) {
  return z.object({
    success: z.literal(true),
    message: z.string(),
    data: dataSchema,
  });
}

export function createListResponseSchema<T extends z.ZodTypeAny>(
  itemSchema: T,
) {
  return z.object({
    success: z.literal(true),
    message: z.string(),
    data: z.array(itemSchema),
    meta: PaginationMetaSchema.optional(),
  });
}

export type ApiSuccessResponse<T = unknown> = {
  success: true;
  message: string;
  data?: T;
  meta?: PaginationMeta;
};
